predictive_test_selection/data_collector.py

```diff
@@ -0,0 +1,236 @@
+        specified in the past CHANGE_LOOK_BACK parameter.
+        """
+        modification_count: List[int] = []
+        for n_days_ago in CHANGE_LOOK_BACK:
+            acc: int = 0
+            for file in self.changed_files:
+                ps = subprocess.Popen(
+                    ["git", "log", f"--after='{n_days_ago} days ago'", "--format=oneline", f"{file}"],
+                    stdout=subprocess.PIPE,
+                )
+                try:
+                    acc += int(subprocess.check_output(["wc", "-l"], stdin=ps.stdout))
+                finally:
+                    ps.stdout.close()
+                    ps.wait()
+
+            modification_count.append(acc)
+        self.modification_count = modification_count
+
+    def _compute_file_extensions(self) -> None:
+        """
+        Computes the set of file extensions present in this code change
+        """
+        self.file_extensions = set([os.path.splitext(file)[1].replace(".", "") for file in self.changed_files])
+
+    def get_modifications_data_format(self) -> str:
+        """
+        Converts the modification_count (List[int]) to a condensed csv format.
+        e.g [1, 2, 3] becomes "1,2,3"
+        """
+        return str(self.modification_count).replace(" ", "")[1:-1]
+
+    def get_file_extensions_data_format(self) -> str:
+        """
+        Converts the file_extensions (Set[str]) to a condensed csv format.
+        e.g {'txt', 'md', 'py'} becomes 'txt,md,py'
+        """
+        return str(self.file_extensions).replace(" ", "").replace("'", "").replace('"', "")[1:-1]
+
+
+@dataclass
+class TestResult:
+    """
+    Collects all data pertaining to the test cases by parsing the test result xml file (produced by
+    '$ py.test --junit-xml=junit-py39.xml'):
+    - test_result_data -> Maps the fully qualified test name <test_file_name>.<test_case> to an integer denoting whether
+      this test failed or not.
+    """
+
+    test_result_data: MutableMapping[str, int] = field(init=False)
+
+    def parse(self) -> None:
+        self.test_result_data = {}
+        path: str = "junit-py39.xml"
+        tree = ET.parse(path)
+        root = tree.getroot()
+
+        for test_suite in root.findall("testsuite"):
+            for test_case in test_suite.findall("testcase"):
+                if test_case.find("skipped") is not None:  # test got skipped -> we don't record anything
+                    continue
+
+                test_file: str = test_case.get("classname")
+                test_name: str = test_case.get("name")
+                test_fqn: str = ".".join((test_file, test_name))
+
+                test_failed: int = int(test_case.find("failure") is not None)
+
+                self.test_result_data[test_fqn] = test_failed
+
+        assert self.test_result_data
+
+    def __iter__(self):
+        yield from self.test_result_data.items()
+
+
+@dataclass
```

Comment:
Why is this a dataclass?

Reply:
No valid reason to use a dataclass there, removing it.